analysis/plot.routes.all_region_pairs.py

```diff
@@ -0,0 +1,196 @@
+import pandas as pd
+import matplotlib.pyplot as plt
+from geopy.distance import geodesic
+from typing import Any, Callable
+
+from common import DirType, init_logging
+
+DATA_SOURCE = 'caida.itdk'
+
+
+def load_weighted_hops(file_path: str) -> pd.DataFrame:
+    """Load the file that contains a list of weights and hops.
+
+        File format:
+        4 hop1|hop2|hop3
+        1 hop1|hop2
+        ...
+    """
+    with open(file_path, 'r') as file:
+        lines = file.readlines()
+
+    # Split each line into two parts
+    l_weight = []
+    l_hops = []
+    for line in lines:
+        parts = line.strip().split(' ', 1)
+        weight = float(parts[0])
+        hops = parts[1]
+        l_weight.append(weight)
+        l_hops.append(hops)
+
+    # Create a DataFrame
+    return pd.DataFrame.from_dict({'weight': l_weight, 'hops': l_hops})
+
+def calculate_total_distance(hops: list[str]) -> float:
+    """Calculate the total distance across all hops.
+
+        Each hop is represented as a string of the form: '(lat,lon)'
+    """
+    if len(hops) <= 1:
+        return 0.
+
+    @functools.cache
+    def calculate_pairwise_distance_km(hop1: str, hop2: str) -> float:
+        (lat1, lon1) = ast.literal_eval(hop1)
+        (lat2, lon2) = ast.literal_eval(hop2)
+        return geodesic((lat1, lon1), (lat2, lon2)).km
+
+    total_distance = 0.
+    for i in range(len(hops) - 1):
+        total_distance += calculate_pairwise_distance_km(hops[i], hops[i + 1])
+    return total_distance
+
+def get_hops_and_weights(file_path: str) -> tuple[np.ndarray, np.ndarray]:
+    data = load_weighted_hops(file_path)
+    hops = data['hops'].to_numpy()
+    weights = data['weight'].to_numpy()
+    return hops, weights
+
+def extract_regions_from_filename(filename):
+    """
+    Extract the source and destination regions from the given filename.
+    Assumes the format: 'routes.[src].[dst].by_iso.distribution'
+    """
+    regex = re.compile(r'^routes\.([a-z0-9-]+)\.([a-z0-9-]+)\.([a-z0-9-]+)\.([a-z0-9-]+)\.by_(?:\w+)\.distribution$')
+    m = regex.match(filename)
+    assert m is not None, 'Invalid filename: {}'.format(filename)
+    src_cloud = m.group(1)
+    src_region = m.group(2)
+    dst_cloud = m.group(3)
+    dst_region = m.group(4)
+    src = f'{src_cloud}:{src_region}'
+    dst = f'{dst_cloud}:{dst_region}'
+    return src, dst
+
+def plot_pdf(values, weights, src_region: str, dst_region: str, metric: str, data_source: str):
+    # Normalize the weights to ensure they sum up to 1
+    weights = np.array(weights) / np.sum(weights)
+
+    plt.hist(values, bins=len(values), weights=weights, density=True, edgecolor='black')
+    plt.xlabel('Values')
+    plt.ylabel('Probability')
+    plt.title(f'PDF of average {metric} between regions {src_region} and {dst_region}')
+
+    filename = f'{metric}.pdf.{data_source}.{src_region}.{dst_region}.png'
+    logging.info(f'Saving heatmap to {filename} ...')
+    plt.savefig(filename, bbox_inches='tight')
```

Comment:
You may also want to call `plt.close()` after you're finished plotting.

Reply:
Thanks! Updating.

Reply:
Actually I don't think it's necessary here. The doc says "close a figure window", and since I'm not opening/showing a figure, I don't think I should call it here.